extensions/common/vault/vault-gcp/src/main/java/org/eclipse/edc/vault/gcp/GcpSecretManagerVault.java

```diff
@@ -0,0 +1,252 @@
+     * @param project GCP project name.
+     * @param region replica location for secrects created by the vault.
+     * @param secretClient GCP client to Secret Manager service (already authenticated).
+     */
+    public GcpSecretManagerVault(Monitor monitor, String project, String region, SecretManagerServiceClient secretClient) {
+        this.monitor = monitor;
+        this.project = project;
+        this.region = region;
+        this.secretManagerServiceClient = secretClient;
+        this.vaultLock = new Object();
+    }
+
+    /**
+     * Retrieves the secret stored under a specific key.
+     *
+     * @param key string key identifying the secret to be fetched.
+     * @return the string if the secret is found, null otherwise.
+     */
+    @Override
+    public @Nullable String resolveSecret(String key) {
+        try {
+            key = sanitizeKey(key);
+            SecretVersionName secretVersionName = SecretVersionName.of(project, key, LATEST_VERSION_ALIAS);
+            AccessSecretVersionResponse response;
+            synchronized (vaultLock) {
+                response = secretManagerServiceClient.accessSecretVersion(secretVersionName);
+            }
+            String payload = response.getPayload().getData().toStringUtf8();
+            return payload;
+        } catch (NotFoundException nfex) {
+            monitor.debug("Secret not found or has no version: " + key);
+            return null;
+        } catch (RuntimeException rex) {
+            monitor.severe("Runtime error while resolving secret " + key + ": " + rex.getMessage(), rex);
+            return null;
+        } catch (Exception ex) {
+            monitor.debug("Exception while resolving secret " + key + ": " + ex.getMessage(), ex);
+            return null;
+        }
+    }
+
+    /**
+     * Saves a secret stored under a specific key. If the selected key is already in use by a secrect, error is
+     * returned and existing secret not overwritten.
+     *
+     * @param key string key identifying the secret to be stored.
+     * @param value string value of the secret.
+     * @return Result.success if the secret is stored.
+     */
+    @Override
+    public Result<Void> storeSecret(String key, String value) {
+        try {
+            key = sanitizeKey(key);
+            Secret secret =
+                    Secret.newBuilder()
+                    .setReplication(
+                        Replication.newBuilder()
+                        .setUserManaged(Replication.UserManaged.newBuilder()
+                            // TODO add multi-region replica support?
+                            .addReplicas(Replication.UserManaged.Replica.newBuilder()
+                                .setLocation(region)
+                                .build())
+                            .build())
+                        .build())
+                    .build();
+
+            ProjectName parent = ProjectName.of(project);
+            synchronized (vaultLock) {
```

Comment:
is the lock mechanism really needed here? What potential issues are we trying to avoid?

Reply:
Likely not needed, and can be removed, will further test as I am not sure yet about how state machine and threads are working. So basically, being cautious :)

Reply:
I guess secret manager is already able to handle concurrency, handling it here seems a bit over-engineered

Reply:
GCP operations are consistent per se, some exceptions are found in the docs (e.g. accessing a secret by alias). What I am missing, while learning about EDC, is how for instance a call to create a secret in one thread is synced to a call for reading the secret in another thread. For this reason, the synchronized block still makes sense to me, I prefer to keep it.

Reply:
the EDC won't ever create the same secret twice from different threads, this eventually would be an EDC issue to fix, not a vault issue. This is too defensive.